Source/WebCore/page/text-extraction/TextExtraction.cpp

```diff
@@ -102,92 +215,183 @@ static inline FloatRect rootViewBounds(Node& node)
-        auto nodeBoundsInRootView = rootViewBounds(node);
-        if (!rectInRootView || rectInRootView->intersects(nodeBoundsInRootView))
-            item = { { WTFMove(*itemData), WTFMove(nodeBoundsInRootView), { } } };
+    std::optional<Editable> editable;
+    std::optional<URL> linkURL;
+
+    WTF::switchOn(extractItemData(node, context),
+        [&](std::monostate) { },
+        [&](URL&& result) { linkURL = WTFMove(result); },
+        [&](Editable&& result) { editable = WTFMove(result); },
+        [&](ItemData&& result) {
+            auto bounds = rootViewBounds(node);
+            if (context.shouldIncludeNodeWithRect(bounds))
+                item = { { WTFMove(result), WTFMove(bounds), { } } };
+        });
+
+    bool onlyCollectTextAndLinks = linkURL || editable;
+    if (onlyCollectTextAndLinks) {
+        if (auto bounds = rootViewBounds(node); context.shouldIncludeNodeWithRect(bounds)) {
+            item = {
+                TextItemData { { }, { }, emptyString(), { } },
+                WTFMove(bounds),
+                { }
+            };
+        }
+        context.onlyCollectTextAndLinksCount++;
+    }
+
+    if (RefPtr container = dynamicDowncast<ContainerNode>(node)) {
+        for (auto& child : composedTreeChildren(*container))
+            extractRecursive(child, item ? *item : parentItem, context);
     }
 
-    if (!item || shouldIncludeChildren(item->data)) {
-        if (RefPtr container = dynamicDowncast<ContainerNode>(node)) {
-            for (auto& child : composedTreeChildren(*container))
-                extractRecursive(child, item ? *item : parentItem, extractedText, rectInRootView);
+    if (onlyCollectTextAndLinks) {
+        if (item) {
+            if (linkURL) {
+                auto& text = std::get<TextItemData>(item->data);
+                text.links.append({ WTFMove(*linkURL), CharacterRange { 0, text.content.length() } });
+            }
+            if (editable) {
+                auto& text = std::get<TextItemData>(item->data);
+                text.editable = WTFMove(editable);
+            }
         }
+        context.onlyCollectTextAndLinksCount--;
     }
 
-    if (item)
-        parentItem.children.append(WTFMove(*item));
+    if (!item)
+        return;
+
+    if (parentItem.children.isEmpty()) {
+        if (canMerge(parentItem, *item))
+            return merge(parentItem, WTFMove(*item));
+    } else if (auto& lastChild = parentItem.children.last(); canMerge(lastChild, *item))
+        return merge(lastChild, WTFMove(*item));
+
+    parentItem.children.append(WTFMove(*item));
+}
+
+static void pruneRedundantItemsRecursive(Item& item)
+{
+    item.children.removeAllMatching([](auto& child) {
+        if (!child.children.isEmpty() || !std::holds_alternative<TextItemData>(child.data))
+            return false;
+
+        auto& text = std::get<TextItemData>(child.data);
+        return !text.editable && text.content.template containsOnly<isASCIIWhitespace>();
```

Comment:
🤯

Reply:
Yeah, it's a bit funky... 😓

It's not actually totally clear why this is necessary, but I did find that if I moved it out of the surrounding lambda, it compiled without the `template`.